ipa-core/src/protocol/context/mod.rs

```diff
@@ -519,21 +509,65 @@ where
     // This approach makes sure we do what we can - send or receive.
     let mut send_recv = pin!(futures::stream::select(send_stream, rcv_stream));
 
-    while let Some((shard_id, v)) = send_recv.next().await {
-        if let Some(m) = v? {
+    while let Some((shard_id, v)) = send_recv.try_next().await? {
+        if let Some(m) = v {
             r[usize::from(shard_id)].push(m);
         }
     }
 
     Ok(r.into_iter().flatten().collect())
 }
 
+/// Provides the same functionality as [`reshard_try_stream`] on
+/// infallible streams
+///
+/// ## Stream size
+/// Note that it currently works for streams where size is known in advance. Mainly because
+/// we want to set up send buffer sizes and avoid sending records one-by-one to each shard.
+/// Other than that, there are no technical limitation here, and it could be possible to make it
+/// work with regular streams or opt-out to runtime checks as [`reshard_try_stream`] does.
+///
+///
+/// ```compile_fail
+/// use futures::stream::{self, StreamExt};
+/// use ipa_core::protocol::context::reshard_stream;
+/// use ipa_core::ff::boolean::Boolean;
+/// use ipa_core::secret_sharing::SharedValue;
+/// async {
+///     let a = [Boolean::ZERO];
+///     let mut s = stream::iter(a.into_iter()).cycle();
+///     // this should fail to compile:
+///     // the trait bound `futures::stream::Cycle<...>: ExactSizeStream` is not satisfied
+///     reshard_stream(todo!(), s, todo!()).await;
+/// };
+/// ```
+/// ## Panics
+/// When `shard_picker` returns an out-of-bounds index.
+///
+/// ## Errors
+/// If cross-shard communication fails
+pub async fn reshard_stream<L, K, C, S>(
+    ctx: C,
+    input: L,
+    shard_picker: S,
+) -> Result<Vec<K>, crate::error::Error>
+where
+    L: ExactSizeStream<Item = K>,
+    S: Fn(C, RecordId, &K) -> ShardIndex,
+    K: Message + Clone,
+    C: ShardedContext,
+{
+    reshard_try_stream(ctx, input.map(Ok), shard_picker).await
```

Comment:
is awaiting here going to cause this to block until the whole stream is processed? wouldn't we want to await higher up the stack?